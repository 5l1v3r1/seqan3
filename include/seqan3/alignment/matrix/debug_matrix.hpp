// -----------------------------------------------------------------------------------------------------
// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
// -----------------------------------------------------------------------------------------------------

/*!\file
 * \author Marcel Ehrhardt <marcel.ehrhardt AT fu-berlin.de>
 * \brief Contains the declaration of seqan3::detail::debug_matrix.
 */

#pragma once

#include <iomanip>

#include <seqan3/alignment/matrix/matrix_concept.hpp>
#include <seqan3/alignment/matrix/row_wise_matrix.hpp>
#include <seqan3/alignment/matrix/trace_directions.hpp>
#include <seqan3/core/debug_stream.hpp>

namespace seqan3::detail
{

/*!\brief A debug matrix to wrap alignment matrices and sequences and make them printable together.
 * \ingroup alignment_matrix
 * \implements seqan3::detail::Matrix
 * \tparam matrix_t    An alignment matrix; Must model seqan3::detail::Matrix.
 * \tparam sequence1_t The type of the first sequence; If no sequences are given this is std::nullopt_t.
 * \tparam sequence2_t The type of the second sequence; If no sequences are given this is std::nullopt_t.
 *
 *
 * This debug matrix allows you to print an alignment matrix (e.g. score or trace matrix) combined with two sequences.
 *
 * \cond DEV
 * This type is used internally
 *   * to print seqan3::detail::Matrix's and,
 *   * to compare alignment matrices in in test cases.
 * \endcond
 *
 * # Score Matrix Example
 *
 * \snippet test/snippet/alignment/matrix/debug_matrix.cpp score_matrix
 *
 * ### Output
 * \snippet test/snippet/alignment/matrix/debug_matrix.out score_matrix::out
 *
 * # Trace Matrix Example
 *
 * \snippet test/snippet/alignment/matrix/debug_matrix.cpp trace_matrix
 *
 * ### Output
 * \snippet test/snippet/alignment/matrix/debug_matrix.out trace_matrix::out
 */
template <Matrix matrix_t, typename sequence1_t = std::nullopt_t, typename sequence2_t = std::nullopt_t>
class debug_matrix
{
public:
    //!\copydoc seqan3::detail::Matrix::entry_type
    using entry_type = typename std::remove_reference_t<matrix_t>::entry_type;

    /*!\name Constructors, destructor and assignment
     * \{
     */
    debug_matrix() = default; //!< Defaulted
    debug_matrix(debug_matrix const &) = default; //!< Defaulted
    debug_matrix(debug_matrix &&) = default; //!< Defaulted
    debug_matrix & operator=(debug_matrix const &) = default; //!< Defaulted
    debug_matrix & operator=(debug_matrix &&) = default; //!< Defaulted
    ~debug_matrix() = default;  //!< Defaulted

    /*!\brief Construct the matrix out of the *entries*, the *rows*, and the *cols*.
     * \param entries The entry values as a flat std::vector <#entry_type>.
     * \param rows    The number of rows.
     * \param cols    The number of columns.
     */
    debug_matrix(std::vector<entry_type> entries, size_t const rows, size_t const cols)
        : _matrix{std::move(entries), rows, cols},
          _sequence1{std::nullopt},
          _sequence2{std::nullopt}
    {}

    /*!\brief Construct the matrix out of the *entries*, the *rows*, and the *cols* and the corresponding sequences.
     * \param entries The entry values as a flat std::vector <#entry_type>.
     * \param rows    The number of rows.
     * \param cols    The number of columns.
     * \param sequence1 The first sequence of the sequence alignment.
     * \param sequence2 The second sequence of the sequence alignment.
     */
    debug_matrix(std::vector<entry_type> entries, size_t const rows, size_t const cols,
                 sequence1_t sequence1, sequence2_t sequence2)
        : _matrix{std::move(entries), rows, cols},
          _sequence1{std::forward<sequence1_t>(sequence1)},
          _sequence2{std::forward<sequence2_t>(sequence2)}
    {
        assert(cols == _sequence1.size() + 1u);
        assert(rows == _sequence2.size() + 1u);
    }

    /*!\brief Construct the matrix out of an existing matrix.
     * \param matrix An alignment matrix; Must model seqan3::detail::Matrix.
     */
    debug_matrix(matrix_t matrix)
        : _matrix{std::forward<matrix_t>(matrix)},
          _sequence1{std::nullopt},
          _sequence2{std::nullopt}
    {}

    /*!\brief Construct the matrix out of an existing matrix and two sequences.
     * \param matrix An alignment matrix; Must model seqan3::detail::Matrix.
     * \param sequence1 The first sequence of the sequence alignment.
     * \param sequence2 The second sequence of the sequence alignment.
     */
    debug_matrix(matrix_t matrix, sequence1_t sequence1, sequence2_t sequence2)
        : _matrix{std::forward<matrix_t>(matrix)},
          _sequence1{std::forward<sequence1_t>(sequence1)},
          _sequence2{std::forward<sequence2_t>(sequence2)}
    {
        assert(_matrix.cols() == _sequence1.size() + 1u);
        assert(_matrix.rows() == _sequence2.size() + 1u);
    }
    //!\}

    //!\copydoc seqan3::detail::Matrix::rows
    size_t rows() const noexcept
    {
        return _matrix.rows();
    }

    //!\copydoc seqan3::detail::Matrix::cols
    size_t cols() const noexcept
    {
        return _matrix.cols();
    }

    //!\copydoc seqan3::detail::debug_matrix::_sequence1
    sequence1_t const & sequence1() const noexcept
    {
        return _sequence1;
    }

    //!\copydoc seqan3::detail::debug_matrix::_sequence2
    sequence2_t const & sequence2() const noexcept
    {
        return _sequence2;
    }

    //!\copydoc seqan3::detail::Matrix::at
    entry_type at(size_t const row, size_t const col) const noexcept
    {
        assert(row < rows() && col < cols());
        return _matrix.at(row, col);
    }

protected:
    //!\cond
    struct format_type; // forward declaration
    //!\endcond

public:
    /*!\brief Prints this matrix into the given stream.
     * \param cout  The stream to print to.
     * \param flags Modify the way the matrix is printed.
     *
     *
     * The matrix will be printed out with unicode characters if seqan3::fmtflags2::utf8 is set in the flags. Ascii
     * otherwise.
     */
    template <typename oostream_t>
    void print(oostream_t & cout, fmtflags2 const flags) const noexcept
    {
        format_type const & symbols = (flags & fmtflags2::utf8) == fmtflags2::utf8 ? unicode : csv;
        size_t const column_width = this->column_width.has_value() ?
                                    this->column_width.value() : auto_column_width(flags);

        auto char_sequence1 = [&]([[maybe_unused]] size_t const i) -> std::string
        {
            if constexpr(std::is_same_v<sequence1_t, std::nullopt_t>)
                return " ";
            else
                return as_string(sequence1()[i], flags);
        };

        auto char_sequence2 = [&]([[maybe_unused]] size_t const i) -> std::string
        {
            if constexpr(std::is_same_v<sequence2_t, std::nullopt_t>)
                return " ";
            else
                return as_string(sequence2()[i], flags);
        };

        auto print_cell = [&](std::string const & symbol)
        {
            // deal with unicode chars that mess up std::setw
            size_t const length_bytes = symbol.size();
            size_t const length = unicode_str_length(symbol);
            size_t const offset = length_bytes - length;

            cout << std::left
                 << std::setw(column_width + offset)
                 << symbol
                 << symbols.col_sep;
        };

        auto print_first_cell = [&](std::string const & symbol)
        {
            cout << symbol << symbols.col_sep;
        };

        // |_|d|a|t|a|b|a|s|e|
        auto print_first_row = [&]
        {
            print_first_cell(" ");
            print_cell(symbols.epsilon);

            for (size_t col = 0; col < cols() - 1; ++col)
                print_cell(char_sequence1(col));
            cout << "\n";
        };

        // |-|-|-|-|-|-|-|-|-|
        auto print_divider = [&]
        {
            cout << " " << symbols.row_col_sep;
            for (size_t col = 0; col < cols(); ++col)
            {
                for (size_t i = 0; i < column_width; ++i)
                    cout << symbols.row_sep;
                cout << symbols.row_col_sep;
            }
            cout << "\n";
        };

        print_first_row();
        for (size_t row = 0; row < rows(); ++row)
        {
            if (symbols.row_sep[0] != '\0')
                print_divider();

            // one query letter + one row of scores / traces
            if (row == 0)
                print_first_cell(symbols.epsilon);
            else
                print_first_cell(char_sequence2(row - 1));

            for (size_t col = 0; col < cols(); ++col)
                print_cell(entry_at(row, col, flags));
            cout << "\n";
        }
    }

    //!\brief Determines the largest width of all entries in the matrix, e.g. `-152` has width 4.
    size_t auto_column_width(fmtflags2 const flags) const noexcept
    {
        size_t col_width = 1;
        for (size_t row = 0; row < rows(); ++row)
            for (size_t col = 0; col < cols(); ++col)
                col_width = std::max(col_width, unicode_str_length(entry_at(row, col, flags)));
        return col_width;
    }

protected:
    //!\brief Whether the entry_type is trace_directions.
    static constexpr bool is_traceback_matrix = std::is_same_v<std::decay_t<entry_type>, trace_directions>;

    //!\brief Same as at(*row*, *col*), but as string.
    std::string entry_at(size_t const row, size_t const col, fmtflags2 flags) const noexcept
    {
        format_type const & symbols = (flags & fmtflags2::utf8) == fmtflags2::utf8 ? unicode : csv;

        entry_type const & entry = at(row, col);
        if (!is_traceback_matrix && entry == matrix_inf<entry_type>)
            return symbols.inf;
        return as_string(entry, flags);
    }

    //!\brief Convert a value into a std::string.
    template <typename entry_type>
    static std::string as_string(entry_type && entry, fmtflags2 const flags) noexcept
    {
        std::stringstream strstream;
        debug_stream_type stream{strstream};
        stream << flags << entry;
        return strstream.str();
    }

    //!\brief The length of the *str* (traceback symbols are unicode aware).
    //!\sa https://en.wikipedia.org/wiki/UTF-8 for encoding details
    static size_t unicode_str_length(std::string const & str) noexcept
    {
        size_t length = 0u;
        for (auto it = str.cbegin(); it < str.cend(); ++it, ++length)
        {
            uint8_t v = *it;
            if ((v & 0b11100000) == 0b11000000)
                ++it;
            else if ((v & 0b11110000) == 0b11100000)
                it += 2;
            else if ((v & 0b11111000) == 0b11110000)
                it += 3;
        }
        return length;
    }

    //!\brief Format used by seqan3::detail::debug_matrix.
    struct format_type
    {
        //!\brief epsilon symbol (a single symbol)
        char const * epsilon{};
        //!\brief column separator symbol (a single symbol)
        char const * col_sep{};
        //!\brief row separator symbol (a single symbol)
        char const * row_sep{};
        //!\brief row separator symbol (a single symbol)
        char const * row_col_sep{};
        //!\brief infinity symbol (a single symbol)
        char const * inf{};
    };

    //!\brief The format when printing to a ascii stream.
    static constexpr format_type csv{" ", ";", "", "", ""};
    //!\brief The format when printing to a unicode stream.
    static constexpr format_type unicode{"ε", "║", "═", "╬", "∞"};

public:
    //!\brief What is the width (number of chars) of an entry. Defaults to #auto_column_width.
    std::optional<size_t> column_width{std::nullopt};

protected:
    //!\brief The matrix
    matrix_t _matrix;
    //!\brief The first sequence of the sequence alignment.
    sequence1_t _sequence1;
    //!\brief The second sequence of the sequence alignment.
    sequence2_t _sequence2;
};

/*!\name Type deduction guides
 * \relates seqan3::detail::debug_matrix
 * \{
 */
//!\brief The type deduction guide for the constructor
//!seqan3::detail::debug_matrix(std::vector<entry_type>, size_t const, size_t const)
template <typename entry_type>
debug_matrix(std::vector<entry_type>, size_t const, const size_t)
    -> debug_matrix<row_wise_matrix<entry_type>>;

//!\brief The type deduction guide for the constructor
//!seqan3::detail::debug_matrix(std::vector<entry_type>, size_t const, size_t const, sequence1_t, sequence2_t)
template <typename entry_type, typename sequence1_t, typename sequence2_t>
debug_matrix(std::vector<entry_type>, size_t const, const size_t, sequence1_t &&, sequence2_t &&)
   -> debug_matrix<row_wise_matrix<entry_type>, sequence1_t, sequence2_t>;


//!\brief The type deduction guide for the constructor seqan3::detail::debug_matrix(matrix_t)
template <Matrix matrix_t>
debug_matrix(matrix_t &&)
   -> debug_matrix<matrix_t>;

//!\brief The type deduction guide for the constructor seqan3::detail::debug_matrix(matrix_t, sequence1_t, sequence2_t)
template <Matrix matrix_t, typename sequence1_t, typename sequence2_t>
debug_matrix(matrix_t &&, sequence1_t &&, sequence2_t &&)
   -> debug_matrix<matrix_t, sequence1_t, sequence2_t>;
//!\}

} // namespace seqan3::detail
