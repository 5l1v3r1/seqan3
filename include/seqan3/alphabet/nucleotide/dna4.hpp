#pragma once

#include "../alphabet.hpp"

namespace seqan3
{

struct dna4
{
    // strictly typed enum, unfortunately with scope
    enum struct c_type : uint8_t
    {
        A,
        C,
        G,
        T,
        U = T,
        UNKNOWN = A
    };

    // import into local scope
    static constexpr c_type A{c_type::A};
    static constexpr c_type C{c_type::C};
    static constexpr c_type G{c_type::G};
    static constexpr c_type T{c_type::T};
    static constexpr c_type U{c_type::U};
    static constexpr c_type UNKNOWN{c_type::UNKNOWN};

    // the value
    c_type value;

    // implicit compatibility to inner_type
    constexpr dna4 & operator =(c_type const c)
    {
        value = c;
    }
    constexpr operator c_type() const
    {
        return value;
    }

    // explicit compatibility to char
    explicit constexpr operator char() const
    {
        return to_char();
    }
    constexpr char to_char() const
    {
        return value_to_char[static_cast<uint8_t>(value)];
    }

    constexpr dna4 from_char(char const c)
    {
        value = char_to_value[c];
        return *this;
    }

    // explicit compatibility to integral
    constexpr char to_integral() const
    {
        return static_cast<std::underlying_type_t<c_type>>(value);
    }

    constexpr dna4 from_integral(std::underlying_type_t<c_type> const c)
    {
        value = static_cast<c_type>(c);
        return *this;
    }

    // conversion tables
    static constexpr uint8_t value_size{4};

    static constexpr char value_to_char[value_size]
    {
        'A',
        'C',
        'G',
        'T'
    };

    static constexpr c_type char_to_value[256]
    {
        c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN,
        c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, //0
        c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN,
        c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, //1
        c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN,
        c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, //2
        c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN,
        c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, //3
        //          , A,            B,            C,            D,            E,            F,            G,
        c_type::UNKNOWN, c_type::A,       c_type::UNKNOWN, c_type::C,       c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::G,
        // H,         I,            J,            K,            L,            M,            N,            O,
        c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, //4
        // P,         Q,            R,            S,            T,            U,            V,            W,
        c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::T,       c_type::T,       c_type::UNKNOWN, c_type::UNKNOWN,
        // X,         Y,            Z,
        c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, //5
        //          , a,            b,            c,            d,            e,            f,            g,
        c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::C,       c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::G,
        // h,         i,            j,            k,            l,            m,            n,            o,
        c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, //6
        // P,         Q,            R,            S,            T,            U,            V,            W,
        c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::T,       c_type::T,       c_type::UNKNOWN, c_type::UNKNOWN,
        c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, //7
        c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN,
        c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, //8
        c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN,
        c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, //9
        c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN,
        c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, //10
        c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN,
        c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, //11
        c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN,
        c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, //12
        c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN,
        c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, //13
        c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN,
        c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, //14
        c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN,
        c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN, c_type::UNKNOWN  //15
    };

};

// shall fulfill Alphabet concept
static_assert(alphabet_concept<dna4>);
static_assert(dna4{dna4::A} == dna4{});
static_assert(dna4{dna4::A} == dna4::A);
// static_assert(dna4{'A'} == 'A');
static_assert(static_cast<char>(dna4{dna4::C}) == 'C');
static_assert(dna4{dna4::A} < dna4{dna4::C});

}
