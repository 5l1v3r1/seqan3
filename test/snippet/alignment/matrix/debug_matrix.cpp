#include <iostream>

#include <seqan3/alphabet/nucleotide/dna4.hpp>
#include <seqan3/alignment/matrix/debug_matrix.hpp>
#include <seqan3/core/debug_stream.hpp>
#include <seqan3/range/view/to_char.hpp>

int main()
{
    using namespace seqan3;
{
    debug_stream << "//! [score_matrix::out]" << std::endl;
//! [score_matrix]
using namespace seqan3;
using namespace seqan3::detail;

std::vector<dna4> database = "AACCGGTT"_dna4;
std::vector<dna4> query = "ACGT"_dna4;

debug_matrix score_matrix
{
    std::vector
    {
        -0, -1, -2, -3, -4, -5, -6, -7, -8,
        -1, -0, -1, -2, -3, -4, -5, -6, -7,
        -2, -1, -1, -1, -2, -3, -4, -5, -6,
        -3, -2, -2, -2, -2, -2, -3, -4, -5,
        -4, -3, -3, -3, -3, -3, -3, -3, -4
    },
    5u,
    9u
};

debug_stream << "database:\t" << database << std::endl;
debug_stream << "query:\t\t" << query << std::endl;
debug_stream << std::endl;

debug_stream << "score_matrix: " << score_matrix.cols() << " columns and "
          << score_matrix.rows() << " rows" << std::endl;

// Prints out the matrix in a convenient way
debug_stream << score_matrix << std::endl; // without sequences
debug_stream << debug_matrix{score_matrix, database, query} << std::endl; // with sequences
debug_stream << fmtflags2::utf8 << debug_matrix{score_matrix, database, query}; // as utf8
//! [score_matrix]
    debug_stream << "//! [score_matrix::out]" << std::endl;
    debug_stream.unsetf(fmtflags2::utf8);
}
{
    debug_stream << "//! [trace_matrix::out]" << std::endl;
//! [trace_matrix]
using namespace seqan3;
using namespace seqan3::detail;

std::vector<dna4> database = "AACCGGTT"_dna4;
std::vector<dna4> query = "ACGT"_dna4;

auto N = trace_directions::none;
auto D = trace_directions::diagonal;
auto U = trace_directions::up;
auto L = trace_directions::left;

debug_matrix trace_matrix
{
    std::vector
    {
        N,L,L  ,L  ,L  ,L  ,L  ,L,L  ,
        U,D,D|L,L  ,L  ,L  ,L  ,L,L  ,
        U,U,D  ,D  ,D|L,L  ,L  ,L,L  ,
        U,U,D|U,D|U,D  ,D  ,D|L,L,L  ,
        U,U,D|U,D|U,D|U,D|U,D  ,D,D|L
    },
    5u,
    9u
};

debug_stream << "database:\t" << database << std::endl;
debug_stream << "query:\t\t" << query << std::endl;
debug_stream << std::endl;

debug_stream << "trace_matrix: " << trace_matrix.cols() << " columns and "
          << trace_matrix.rows() << " rows" << std::endl;

// Prints out the matrix in a convenient way
debug_stream << trace_matrix << std::endl; // without sequences
debug_stream << debug_matrix{trace_matrix, database, query} << std::endl; // with sequences
debug_stream << fmtflags2::utf8 << debug_matrix{trace_matrix, database, query}; // as utf8
//! [trace_matrix]
    debug_stream << "//! [trace_matrix::out]" << std::endl;
}
    return 0;
}
