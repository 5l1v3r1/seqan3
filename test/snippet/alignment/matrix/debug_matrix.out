//! [score_matrix::out]
database:	AACCGGTT
query:		ACGT

score_matrix: 9 columns and 5 rows
 ;  ;  ;  ;  ;  ;  ;  ;  ;  ;
 ;0 ;-1;-2;-3;-4;-5;-6;-7;-8;
 ;-1;0 ;-1;-2;-3;-4;-5;-6;-7;
 ;-2;-1;-1;-1;-2;-3;-4;-5;-6;
 ;-3;-2;-2;-2;-2;-2;-3;-4;-5;
 ;-4;-3;-3;-3;-3;-3;-3;-3;-4;

 ;  ;A ;A ;C ;C ;G ;G ;T ;T ;
 ;0 ;-1;-2;-3;-4;-5;-6;-7;-8;
A;-1;0 ;-1;-2;-3;-4;-5;-6;-7;
C;-2;-1;-1;-1;-2;-3;-4;-5;-6;
G;-3;-2;-2;-2;-2;-2;-3;-4;-5;
T;-4;-3;-3;-3;-3;-3;-3;-3;-4;

 ║ε ║A ║A ║C ║C ║G ║G ║T ║T ║
 ╬══╬══╬══╬══╬══╬══╬══╬══╬══╬
ε║0 ║-1║-2║-3║-4║-5║-6║-7║-8║
 ╬══╬══╬══╬══╬══╬══╬══╬══╬══╬
A║-1║0 ║-1║-2║-3║-4║-5║-6║-7║
 ╬══╬══╬══╬══╬══╬══╬══╬══╬══╬
C║-2║-1║-1║-1║-2║-3║-4║-5║-6║
 ╬══╬══╬══╬══╬══╬══╬══╬══╬══╬
G║-3║-2║-2║-2║-2║-2║-3║-4║-5║
 ╬══╬══╬══╬══╬══╬══╬══╬══╬══╬
T║-4║-3║-3║-3║-3║-3║-3║-3║-4║
//! [score_matrix::out]
//! [trace_matrix::out]
database:	AACCGGTT
query:		ACGT

trace_matrix: 9 columns and 5 rows
 ;  ;  ;  ;  ;  ;  ;  ;  ;  ;
 ;N ;L ;L ;L ;L ;L ;L ;L ;L ;
 ;U ;D ;DL;L ;L ;L ;L ;L ;L ;
 ;U ;U ;D ;D ;DL;L ;L ;L ;L ;
 ;U ;U ;DU;DU;D ;D ;DL;L ;L ;
 ;U ;U ;DU;DU;DU;DU;D ;D ;DL;

 ;  ;A ;A ;C ;C ;G ;G ;T ;T ;
 ;N ;L ;L ;L ;L ;L ;L ;L ;L ;
A;U ;D ;DL;L ;L ;L ;L ;L ;L ;
C;U ;U ;D ;D ;DL;L ;L ;L ;L ;
G;U ;U ;DU;DU;D ;D ;DL;L ;L ;
T;U ;U ;DU;DU;DU;DU;D ;D ;DL;

 ║ε ║A ║A ║C ║C ║G ║G ║T ║T ║
 ╬══╬══╬══╬══╬══╬══╬══╬══╬══╬
ε║↺ ║← ║← ║← ║← ║← ║← ║← ║← ║
 ╬══╬══╬══╬══╬══╬══╬══╬══╬══╬
A║↑ ║↖ ║↖←║← ║← ║← ║← ║← ║← ║
 ╬══╬══╬══╬══╬══╬══╬══╬══╬══╬
C║↑ ║↑ ║↖ ║↖ ║↖←║← ║← ║← ║← ║
 ╬══╬══╬══╬══╬══╬══╬══╬══╬══╬
G║↑ ║↑ ║↖↑║↖↑║↖ ║↖ ║↖←║← ║← ║
 ╬══╬══╬══╬══╬══╬══╬══╬══╬══╬
T║↑ ║↑ ║↖↑║↖↑║↖↑║↖↑║↖ ║↖ ║↖←║
//! [trace_matrix::out]
